### PR TITLE
msbuild: init at 16.3

### LINKED ISSUE #29817: mono50 doesn't include msbuild
## Issue description

As of mono 5.0 mono's xbuild tool is being [deprecated in favor of the finally-open-sourced msbuild](http://www.mono-project.com/docs/about-mono/releases/5.0.0/#msbuild). Mono now ships with msbuild.

[Current mono nix 5.0 package](https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/compilers/mono) includes xbuild but msbuild is nowhere to be found in my /nix/store, nor on the PATH.

### Steps to reproduce

(I will improve these)

1. Run `nix-shell -p mono50`
2. `which msbuild`

## Technical details

* System: ProductName: Mac OS X; ProductVersion: 10.12.6; BuildVersion: 16G29
* Nix version: nix-shell (Nix) 1.11.10
* Nixpkgs version: 17.09pre108299.ec9a23332f
* Sandboxing enabled: No

### PR DESCRIPTION
**Pull request by jirkadanek**
###### Motivation for this change
Based upon https://github.com/NixOS/nixpkgs/pull/43680 and https://github.com/NixOS/nixpkgs/pull/50842
(Sort of) fixes https://github.com/NixOS/nixpkgs/issues/29817
Depends on PRs https://github.com/NixOS/nixpkgs/pull/65875 and https://github.com/NixOS/nixpkgs/pull/66131

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @thoughtpolice @obadz @vrthra @worldofpeace @samueldr

### CODE REVIEW

**Comment by jdanekrh** [comment]
I guess I did all I could at this point. I can build and run msbuild on my machine.

Known issues: My checkPhase does not seem to be running, I am skipping tests during build, and I am getting a failure from the ci bot which I don't understand.

**Comment by worldofpeace** [comment]
Have you tried `doCheck = true` @jdanekrh?

**Comment by jdanekrh** [comment]
Rebased and ready for review. Only outstanding work to do is the disabled unit tests during build. Can I get it in without them? I have installCheckPhase, at least.

**Comment by jdanekrh** [comment]
One more issue. Jetbrains.rider does not accept the wrapped executable, it only takes the dll. But then the MSBuildExtensionsPath variable is not configured and builds fail. I can fix that in the rider gui by setting msbuild global property, but it is a hassle.

**Comment by jdanekrh** [comment]
Still one more MSBuildExtensionsPath thing. OmniSharp autoconfiguration does not pick it correctly either, see https://github.com/NixOS/nixpkgs/pull/43680#issuecomment-518622169.

**Comment by Mic92** [comment]
@jdanekrh maybe you can have an unwrapped version that is passed to rider and then you set MSBuildExtensionsPath in rider itself.

**Comment by jdanekrh** [comment]
@Mic92 The wrapper must be a shell script, though. Right. I cannot wrap a .net dll in another dll.

Rider needs to be given a dll, if I give it a shell script, it screams

>  Cannot open assembly '/nix/store/sn4h4a3qbp11yng2621yqzk16blvn6bw-msbuild-16.3+xamarinxplat.2019.07.26.14.57/bin/msbuild': File does not contain a valid CIL image.

**Comment by jdanekrh** [comment]
Sorry, I misunderstood the comment. Yes, you are absolutely right. What you said is what I am doing to make this work for me.

File: pkgs/development/tools/build-managers/msbuild/nuget.nix [hunk 1]
```diff
@@ -0,0 +1,1130 @@
+{ fetchurl }: let
+
+  fetchNuGet = { url, name, version, sha256 }: fetchurl {
+    inherit name url sha256;
+  };
+
+in [
+(fetchNuGet {
```
Comment (Mic92) [inline pkgs/development/tools/build-managers/msbuild/nuget.nix]: Is this generated?

Comment (jdanekrh) [reply pkgs/development/tools/build-managers/msbuild/nuget.nix]: Yes, I adapted what https://github.com/NixOS/nixpkgs/tree/master/pkgs/servers/nosql/eventstore does.

**Comment by jdanekrh** [comment]
This was merged as part of PR #66299 in commit a9b4e7592faad4852236245d333ea72c2cb282b8

**Pull request closed without merging**